### PR TITLE
Flatiron Jenkins continuous integration

### PR DESCRIPTION
**Pull request by lyndond**
With Dylan's help, we are setting up GPU testing w/ Jenkins at Flatiron Inst. 

- All pytest files run green on both gpu and cpu

### CODE REVIEW

File: jenkins/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        HOME = pwd(tmp:true)
+	OMP_NUM_THREADS = 4
+      }
+      steps {
+	sh 'python3 -c "import torch; torch.cuda.current_device()"'
+	sh 'python3 -m venv --system-site-packages $HOME'
+	sh '''#!/bin/bash -ex
+	  source $HOME/bin/activate
+	  pip3 install .
+      for f in tests/test_*py; do python3 -m pytest $f; done
```
Comment (billbrod) [inline jenkins/Jenkinsfile:26]: probably worth adding a comment here to say why we do it this way instead of the obvious `python3 -m pytest tests`

File: plenoptic/metric/perceptual_distance.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def _ssim_parts(img1, img2, dynamic_range):
 
     real_size = min(11, height, width)
     std = torch.tensor(1.5).to(img1.device)
-    window = circular_gaussian2d(real_size, std=std, out_channels=n_channels)
+    window = circular_gaussian2d(real_size, std, n_channels, img1.device)
```
Comment (billbrod) [inline plenoptic/metric/perceptual_distance.py:69]: is adding a device arg better than calling `.to(device)` on the resulting tensor?

Comment (billbrod) [reply plenoptic/metric/perceptual_distance.py:69]: especially because it doesn't look like we're consistent: `models.naive.Gaussian` doesn't accept a device arg. Is the idea that models don't accept a device arg, but components can?

File: plenoptic/simulate/models/naive.py [hunk 3]
```diff
@@ -136,7 +136,9 @@ def filt(self):
         if self._filt is not None:  # use old filter
             return self._filt
         else:  # create new filter, optionally cache it
-            filt = circular_gaussian2d(self.kernel_size, self.std, self.out_channels)
+            device = self.std.device
+            filt = circular_gaussian2d(self.kernel_size, self.std, self.out_channels, device)
+            filt = filt.to(device)
```
Comment (billbrod) [inline plenoptic/simulate/models/naive.py:141]: is the `filt.to(device)` necessary with the device arg passed to `circular_gaussian2d`?

File: plenoptic/simulate/models/naive.py [hunk 4]
```diff
@@ -235,27 +237,28 @@ def __init__(
 
         self.cache_filt = cache_filt
         self._filt = None
+        self._input_device = torch.device("cpu")
```
Comment (billbrod) [inline plenoptic/simulate/models/naive.py:240]: I think this is unused, yeah? in which case, should delete it

File: plenoptic/simulate/models/naive.py [hunk 5]
```diff
@@ -235,27 +237,28 @@ def __init__(
...
+            filt_surround = circular_gaussian2d(self.kernel_size, self.surround_std, self.out_channels, device)
 
             # sign is + or - depending on center is on or off
-            sign = torch.as_tensor([1. if x else -1. for x in self.on_center])
+            sign = torch.as_tensor([1. if x else -1. for x in self.on_center]).to(device)
             sign = sign.view(self.out_channels, 1, 1, 1)
-            filt = on_amp * (sign * (filt_center - filt_surround))
 
-            filt = filt / filt.sum()
+            filt = on_amp * (sign * (filt_center - filt_surround))
```
Comment (billbrod) [inline plenoptic/simulate/models/naive.py:258]: why did we get rid of  the normalization of the filter?

File: plenoptic/synthesize/eigendistortion.py [hunk 6]
```diff
@@ -282,7 +282,7 @@ def _synthesize_exact(self) -> Tuple[Tensor, Tensor]:
 
         J = self.compute_jacobian()
         F = J.T @ J
-        eig_vals, eig_vecs = torch.symeig(F, eigenvectors=True)
+        eig_vals, eig_vecs = torch.linalg.eigh(F, UPLO="U")
```
Comment (billbrod) [inline plenoptic/synthesize/eigendistortion.py:285]: isn't `torch.linalg` only available starting in `torch` 1.10?

Comment (billbrod) [reply plenoptic/synthesize/eigendistortion.py:285]: Also, this seems unrelated to the Jenkins / GPU tests issue, and so probably shouldn't be part of this PR?

File: tests/test_eigendistortion.py [hunk 7]
```diff
@@ -139,17 +139,16 @@ def test_display(self, model, einstein_img, color_img, method, k):
...
 class TestAutodiffFunctions:
 
     @pytest.fixture(scope='class')
     def state(self, einstein_img):
         """variables to be reused across tests in this class"""
-        torch.manual_seed(0)
 
         k = 2  # num vectors with which to compute vjp, jvp, Fv
-        einstein_img = einstein_img[..., :16, :16]  # reduce image size
+        einstein_img = einstein_img[..., 100:100+16, 100:100+16]  # reduce image size
```
Comment (billbrod) [inline tests/test_eigendistortion.py:151]: why grab a different part of the image?

File: tests/test_eigendistortion.py [hunk 8]
```diff
@@ -193,16 +192,20 @@ def test_jac_vec_prod(self, state):
     def test_fisher_vec_prod(self, state):
         x, y, x_dim, y_dim, k = state
 
-        V = torch.randn((x_dim, k), device=DEVICE)
+        V, _ = torch.linalg.qr(torch.ones((x_dim, k), device=DEVICE), "reduced")
+        U = V.clone()
```
Comment (billbrod) [inline tests/test_eigendistortion.py:196]: - same comment about `torch.linalg` and torch 1.10
- why use the cloned version?

**Review by billbrod — CHANGES_REQUESTED** [review]
I've got a couple of small questions to address here:
- when should device be an argument vs should we call `.to(device)`
- you used `torch.linalg` in a couple of places, but I thought that wasn't supported until 1.10
- there are some changes in eigendistortions.py that look unrelated to this PR

File: plenoptic/synthesize/eigendistortion.py (hunk 6 above)
Comment (nikparth) [reply plenoptic/synthesize/eigendistortion.py:285]: I believe torch.linalg.eigh is still there in torch 1.8 which we require so I don't think this is an issue

File: tests/test_eigendistortion.py [hunk 9]
```diff
@@ -193,16 +192,20 @@ def test_jac_vec_prod(self, state):
...
 
-        Fv2 = jac.T @ jac @ V  # manually compute product to compare accuracy
+        Fv2 = jac.T @ jac @ U  # manually compute product to compare accuracy
 
         assert Fv.shape == (x_dim, k)
-        assert Fv2.allclose(Fv, rtol=1E-2)
+        if DEVICE != torch.device("cuda"):
+            assert Fv2.allclose(Fv, rtol=1E-2) 
+        else:
+            assert Fv2.allclose(Fv, atol=1E-6)
```
Comment (nikparth) [inline tests/test_eigendistortion.py:208]: why check rtol for not cuda and atol for cuda? is there not a way to keep these the same just for clarity?

File: tests/test_metric.py [hunk 10]
```diff
@@ -197,20 +198,25 @@ def test_msssim_analysis(self, msssim_images):
     def test_nlpd(self, einstein_img, curie_img):
         curie_img.requires_grad_()
         assert po.metric.nlpd(einstein_img, curie_img).requires_grad
+        curie_img.requires_grad_(False)
 
     def test_nspd(self, einstein_img, curie_img):
         curie_img.requires_grad_()
         assert po.metric.nspd(einstein_img, curie_img).requires_grad
+        curie_img.requires_grad_(False)
```
Comment (nikparth) [inline tests/test_metric.py:206]: why does the requires_grad need to be set to False after setting it to True first?

**Review by nikparth — APPROVED** [review]
had a couple minor questions but nothing really important. Seems good to go otherwise.

File: jenkins/Jenkinsfile (hunk 1 above)
Comment (billbrod) [reply jenkins/Jenkinsfile]: back to `pytest tests`, so this isn't relevant anymore

File: plenoptic/synthesize/eigendistortion.py (hunk 6 above)
Comment (lyndond) [reply plenoptic/synthesize/eigendistortion.py:285]: yeah as nikhil said, it works as of 1.8 so there's no harm

File: tests/test_eigendistortion.py (hunk 7 above)
Comment (lyndond) [reply tests/test_eigendistortion.py:151]: There wasn't much going on in the top corner of the image. Wanted to use a patch with more variation so that the eigenvectors would be better behaved.

File: plenoptic/simulate/models/naive.py (hunk 4 above)
Comment (lyndond) [reply plenoptic/simulate/models/naive.py:240]: done

File: plenoptic/simulate/models/naive.py (hunk 3 above)
Comment (lyndond) [reply plenoptic/simulate/models/naive.py:141]: Good catch. No it is not. Removed.

File: plenoptic/simulate/models/naive.py (hunk 5 above)
Comment (lyndond) [reply plenoptic/simulate/models/naive.py:258]: It didn't really make sense bc the filter contains both pos and neg values, so it isn't clear what dividing by its sum would achieve.

File: tests/test_metric.py (hunk 10 above)
Comment (lyndond) [reply tests/test_metric.py:217]: It's for juggling w/ pytest fixtures. `curie_img` is used in a bunch of tests instead of instantiating new data each test.

File: tests/test_eigendistortion.py (hunk 8 above)
Comment (lyndond) [reply tests/test_eigendistortion.py:196]: - torch.linalg intro'd in 1.8.
- cloned bc i'm checking against values and didn't want to accidentally refer to same underlying object

File: plenoptic/metric/perceptual_distance.py (hunk 2 above)
Comment (lyndond) [reply plenoptic/metric/perceptual_distance.py:73]: removed device arg in favour of better device inference within the `circular_gaussian2d` function

File: tests/test_eigendistortion.py (hunk 9 above)
Comment (lyndond) [reply tests/test_eigendistortion.py]: made them both atol for clarity

**Review by billbrod — APPROVED** [review]
